toolchain/lex/tokenized_buffer.cpp

```diff
@@ -300,8 +300,18 @@ class [[clang::internal_linkage]] TokenizedBuffer::Lexer {
   auto LexHorizontalWhitespace(llvm::StringRef& source_text) -> void {
     CARBON_DCHECK(source_text.front() == ' ' || source_text.front() == '\t');
     NoteWhitespace();
-    ++current_column_;
-    source_text = source_text.drop_front();
+    // Handle adjacent whitespace quickly. This comes up frequently for example
+    // due to indentation. We don't expect *huge* runs, so just use a scalar
+    // loop. While still scalar, this avoids repeated table dispatch and marking
+    // whitespace. We use `ssize_t` in the loop for performance.
+    ssize_t ws_count = 1;
+    ssize_t size = source_text.size();
+    while (ws_count < size &&
+           (source_text[ws_count] == ' ' || source_text[ws_count] == '\t')) {
+      ++ws_count;
+    }
+    current_column_ += ws_count;
+    source_text = source_text.drop_front(ws_count);
```

Comment:
Can we simplify this to
```suggestion
    do {
      ++current_column_;
      source_text = source_text.drop_front();
    } while (!source_text.empty() &&
             (source_text.front() == ' ' || source_text.front() == '\t'));
```
... or does that not optimize as well?

Reply:
Incidentally: would it be worthwhile to track the start of the current line instead of the current column, so we don't need to update it each time we update `source_text`?

Reply:
Sadly, the suggested code won't optimize well. =/ Using word-sized local variables for the loop body is pretty key.

But your other comment is spot on -- the whole update structure is a bit wrong here. That's actually coming up on my queue to rework, but it'll be a follow-up as it's a bit invasive. None of the incrementally updated variables are really structured well and its showing now that everything else has been sped up. Stay tuned!